server/test/test_repository.go

```diff
@@ -0,0 +1,236 @@
+	test.ID = id
+	return test
+}
+
+func (r *repository) Provision(ctx context.Context, test Test) error {
+	return nil
+}
+
+func (r *repository) SortingFields() []string {
+	return []string{"created", "name", "last_run"}
+}
+
+const (
+	getTestSQL = `
+	SELECT
+		t.id,
+		t.version,
+		t.name,
+		t.description,
+		t.service_under_test,
+		t.specs,
+		t.outputs,
+		t.created_at,
+		(SELECT COUNT(*) FROM test_runs tr WHERE tr.test_id = t.id) as total_runs,
+		last_test_run.created_at as last_test_run_time,
+		last_test_run.pass as last_test_run_pass,
+		last_test_run.fail as last_test_run_fail
+	FROM tests t
+	LEFT OUTER JOIN (
+		SELECT MAX(id) as id, test_id FROM test_runs GROUP BY test_id
+	) as ltr ON ltr.test_id = t.id
+	LEFT OUTER JOIN
+		test_runs last_test_run
+	ON last_test_run.test_id = ltr.test_id AND last_test_run.id = ltr.id
+`
+
+	testMaxVersionQuery = `
+	INNER JOIN (
+		SELECT id as idx, max(version) as latest_version FROM tests GROUP BY idx
+		) as latest_tests ON latest_tests.idx = t.id AND t.version = latest_tests.latest_version
+	`
+)
+
+func (r *repository) List(ctx context.Context, take, skip int, query, sortBy, sortDirection string) ([]Test, error) {
+	hasSearchQuery := query != ""
+	cleanSearchQuery := "%" + strings.ReplaceAll(query, " ", "%") + "%"
+	params := []any{take, skip}
+
+	sql := getTestSQL + testMaxVersionQuery
+
+	const condition = " AND (t.name ilike $3 OR t.description ilike $3)"
+	if hasSearchQuery {
+		params = append(params, cleanSearchQuery)
+		sql += condition
+	}
+
+	sortingFields := map[string]string{
+		"created":  "t.created_at",
+		"name":     "t.name",
+		"last_run": "last_test_run_time",
+	}
+
+	sql = sortQuery(sql, sortBy, sortDirection, sortingFields)
+	sql += ` LIMIT $1 OFFSET $2 `
+
+	stmt, err := r.db.Prepare(sql)
+	if err != nil {
+		return []Test{}, err
+	}
+	defer stmt.Close()
+
+	rows, err := stmt.QueryContext(ctx, params...)
+	if err != nil {
+		return []Test{}, err
+	}
+
+	tests, err := r.readTestRows(ctx, rows)
+	if err != nil {
+		return []Test{}, err
+	}
+
+	return tests, nil
+}
+
+func sortQuery(sql, sortBy, sortDirection string, sortingFields map[string]string) string {
+	sortField, ok := sortingFields[sortBy]
+
+	if !ok {
+		sortField = sortingFields["created"]
+	}
+
+	dir := "DESC"
+	if strings.ToLower(sortDirection) == "asc" {
+		dir = "ASC"
+	}
+
+	return fmt.Sprintf("%s ORDER BY %s %s", sql, sortField, dir)
+}
+
+func (r *repository) Count(ctx context.Context, query string) (int, error) {
+	countQuery := "SELECT COUNT(*) FROM tests t" + testMaxVersionQuery
+
+	if query != "" {
+		countQuery = fmt.Sprintf("%s WHERE %s", countQuery, query)
+	}
+
+	count := 0
+
+	err := r.db.
+		QueryRowContext(ctx, countQuery).
+		Scan(&count)
+
+	if err != nil {
+		return 0, fmt.Errorf("sql query: %w", err)
+	}
+
+	return count, nil
+}
+
+type scanner interface {
+	Scan(dest ...interface{}) error
+}
+
+func (r *repository) readTestRows(ctx context.Context, rows *sql.Rows) ([]Test, error) {
```

Comment:
These new repos aren't shared, so we can normalize in the functions not including the entity name
```suggestion
func (r *repository) readRows(ctx context.Context, rows *sql.Rows) ([]Test, error) {
```